src/Compiler/Driver/CompilerImports.fs

```diff
@@ -54,27 +55,57 @@ let (++) x s = x @ [ s ]
 
 let IsSignatureDataResource (r: ILResource) =
     r.Name.StartsWithOrdinal FSharpSignatureDataResourceName
+    || r.Name.StartsWithOrdinal FSharpSignatureCompressDataResourceName
     || r.Name.StartsWithOrdinal FSharpSignatureDataResourceName2
 
 let IsOptimizationDataResource (r: ILResource) =
     r.Name.StartsWithOrdinal FSharpOptimizationDataResourceName
+    || r.Name.StartsWithOrdinal FSharpOptimizationCompressDataResourceName
     || r.Name.StartsWithOrdinal FSharpOptimizationDataResourceName2
 
-let GetSignatureDataResourceName (r: ILResource) =
-    if r.Name.StartsWithOrdinal FSharpSignatureDataResourceName then
-        String.dropPrefix r.Name FSharpSignatureDataResourceName
-    elif r.Name.StartsWithOrdinal FSharpSignatureDataResourceName2 then
-        String.dropPrefix r.Name FSharpSignatureDataResourceName2
+let decompressResource (r: ILResource) =
+    use raw = r.GetBytes().AsStream()
+    use decompressed = new MemoryStream()
+    use deflator = new DeflateStream(raw, CompressionMode.Decompress)
+    deflator.CopyTo decompressed
+    deflator.Close()
+    ByteStorage.FromByteArray(decompressed.ToArray()).GetByteMemory()
+
+let GetResourceNameAndSignatureDataFunc (r: ILResource) =
+    let getResourceTypeAndName (r: ILResource) =
+        if r.Name.StartsWithOrdinal FSharpSignatureDataResourceName then
+            FSharpSignatureDataResourceName, String.dropPrefix r.Name FSharpSignatureDataResourceName
+        elif r.Name.StartsWithOrdinal FSharpSignatureCompressDataResourceName then
+            FSharpSignatureCompressDataResourceName, String.dropPrefix r.Name FSharpSignatureCompressDataResourceName
+        elif r.Name.StartsWithOrdinal FSharpSignatureDataResourceName2 then
+            FSharpSignatureDataResourceName2, String.dropPrefix r.Name FSharpSignatureDataResourceName2
+        else
+            failwith "GetSignatureDataResourceName"
+
+    let resourceType, ccuName = getResourceTypeAndName r
+
+    if resourceType = FSharpSignatureCompressDataResourceName then
+        ccuName, (fun () -> decompressResource (r))
     else
-        failwith "GetSignatureDataResourceName"
+        ccuName, (fun () -> r.GetBytes())
+
+let GetResourceNameAndOptimizationDataFunc (r: ILResource) =
```

Comment:
This has exactly the same code as `GetResourceNameAndSignatureDataFunc`, maybe worth to unity it?

Reply:
Yes, I can crtainly see about that.

Reply:
The code isn't quite exactly the same:  
I.e
first:
````
    let resourceType, ccuName =
        if r.Name.StartsWithOrdinal FSharpSignatureDataResourceName then
            FSharpSignatureDataResourceName, String.dropPrefix r.Name FSharpSignatureDataResourceName
````
second:
````
    let resourceType, ccuName =
        if r.Name.StartsWithOrdinal FSharpOptimizationDataResourceName then
            FSharpOptimizationDataResourceName, String.dropPrefix r.Name FSharpOptimizationDataResourceName
````
and this difference is continued for every path through the func.

I could parameterize it I suppose with the actual string values.  I'm not sure it would it more readable or worth it.

Reply:
Yeah it just caught my eye that both of them have identical logic:
```fsharp
let GetResourceNameAnd<X>DataFunc (r: ILResource) =
    let resourceType, ccuName =
        if r.Name.StartsWithOrdinal FSharp<X>DataResourceName then
            FSharp<X>DataResourceName, String.dropPrefix r.Name FSharp<X>DataResourceName
        elif r.Name.StartsWithOrdinal FSharp<X>CompressedDataResourceName then
            FSharp<X>CompressedDataResourceName, String.dropPrefix r.Name FSharp<X>CompressedDataResourceName
        elif r.Name.StartsWithOrdinal FSharp<X>DataResourceName2 then
            FSharp<X>DataResourceName2, String.dropPrefix r.Name FSharp<X>DataResourceName2
        else
            failwith "Get<X>DataResourceName"

    if resourceType = FSharp<X>CompressedDataResourceName then
        ccuName, (fun () -> decompressResource (r))
    else
        ccuName, (fun () -> r.GetBytes())
```
where `<X>` is `Optimization` or `Signature`

Unfortunately it can't be parametrized with just `<X>`, so might as well leave it alone. It's a bit better now that the inner function isn't there anymore.